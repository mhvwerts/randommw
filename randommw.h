/*==========================================================================
 * randommw.h 
 * combines Doornik's zigrandom.h and zignor.h, with modifications and
 * extensions
 *
 * Werts, 2024
 *==========================================================================*/




/* from zigrandom.h */
#ifndef ZIGRANDOM_H
#define ZIGRANDOM_H

#include <stdint.h>

#define M_RAN_INVM30	9.31322574615478515625e-010			  /* 1.0 / 2^30 */
#define M_RAN_INVM32	2.32830643653869628906e-010			  /* 1.0 / 2^32 */
#define M_RAN_INVM48	3.55271367880050092936e-015			  /* 1.0 / 2^48 */
#define M_RAN_INVM52	2.22044604925031308085e-016			  /* 1.0 / 2^52 */
#define M_RAN_INVM64	5.42101086242752217004e-020			  /* 1.0 / 2^64 */

#define RANDBL_32new(iRan1)                   \
    ((int)(iRan1) * M_RAN_INVM32 + (0.5 + M_RAN_INVM32 / 2))
#define RANDBL_48new(iRan1, iRan2)            \
    ((int)(iRan1) * M_RAN_INVM32 + (0.5 + M_RAN_INVM48 / 2) + \
        (int)((iRan2) & 0x0000FFFF) * M_RAN_INVM48)
#define RANDBL_52new(iRan1, iRan2)            \
    ((int)(iRan1) * M_RAN_INVM32 + (0.5 + M_RAN_INVM52 / 2) + \
        (int)((iRan2) & 0x000FFFFF) * M_RAN_INVM52)

/* xoshiro256+ */
void RanSetSeed_xoshiro256p(uint64_t uSeed);
uint32_t IRan_xoshiro256p(void);
double DRan_xoshiro256p(void);


/* MWC8222 George Marsaglia */

/* void 	GetInitialSeeds(uint32_t auiSeed[], int32_t cSeed,
	                    uint32_t uiSeed, uint32_t uiMin);
*/

void RanSetSeed_MWC8222(uint64_t uSeed);
uint32_t IRan_MWC8222(void);
double DRan_MWC8222(void);
double DRan_MWC_52(void);

/* plug-in RNG */
typedef double 		( * DRANFUN)(void);
typedef uint32_t 	( * IRANFUN)(void);
typedef void   		( * RANSETSEEDFUN)(uint64_t);

void    RanSetRan(const char *sRan);
void    RanSetRanExt(DRANFUN DRanFun, IRANFUN IRanFun, RANSETSEEDFUN RanSetSeedFun);
double  DRanU(void);
uint32_t  IRanU(void);
void    RanSetSeed(uint64_t uSeed);

/* normal probabilities */
double  DProbNormal(double x);

/* polar standard normal RNG */
double  DRanNormalPolar(void);
double  FRanQuanNormal(void);
double  DRanQuanNormal(void);

#endif /* ZIGRANDOM_H */


/* from zignor.h */
double  DRanNormalZig(void);
double  DRanQuanNormalZig(void);

/* xoshiro256 */

/* further extensions */
void  RanInit(uint64_t uSeed);
