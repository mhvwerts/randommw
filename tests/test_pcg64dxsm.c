/*
 Run this C program and compare its output to the output generated by 
 the Python/Numpy script `test_pcg64dxsm_numpy_compare.py`. It will 
 (should) produce exactly the same sequences of random integers, also after 
 "jumping". 
 
 This demonstrates that the PCG64DXSM RNG in `randommw.c` is
 working correctly and in conformity to the PCG64DXSM RNG in Numpy.

 The PCG64DXSM initialization code in randommw.c has been modified
 to reproduce exactly the state of the PCG64DXSM RNG obtained by 
 numpy.random.PGC64DXSM(seed = 0) and PGC64DXSM(seed = 12345).
 This is not automatic, since the initialization of PCG64DXSM is not the 
 same in randommw.c and in Numpy.  It is not feasible to use exactly the 
 same initialization of the PCG64DXSM in `randommw.c` as in Numpy. The 
 latter depends on the specific Numpy infrastructure. In `randommw.c`, 
 SplitMix64 is used for initialization.

*/

#include <stdint.h>
#include <inttypes.h>
#include <stdio.h>

#include "randommw.h"

extern void pcg64dxsm_getstateinc(uint64_t *dataptr);
extern uint64_t pcg64dxsm_next();

static void printstate()
{
	uint64_t stateinc[4]; // to store state/inc of pcg64dxsm rng for diagnostic
	
	pcg64dxsm_getstateinc(stateinc);
	printf("state: 0x%016"PRIx64 "%016" PRIx64 "\n", stateinc[0], stateinc[1]);
	printf("inc:   0x%016"PRIx64 "%016" PRIx64 "\n", stateinc[2], stateinc[3]);
}

static void starline()
{
	printf("************************************************************\n");
}


int main(void)
{


	
	uint64_t sseed[2] = {0, 12345};
	uint64_t Njmp[2]     = {7, 64738};

	uint64_t jj, ii;


	for(jj = 0; jj<2; jj++)
	{
		printf("\n");
		starline();
		printf("*** Seed : %"PRIu64"\n", sseed[jj]);
		RanInit("PCG64DXSM", sseed[jj], 0);
		starline();

		printstate();
		printf("\n");
		for (ii = 0; ii < 10; ii++)
		{
			printf("%"PRIu64"\n", pcg64dxsm_next());
		}
		printf("\n");
		printstate();

			
		printf("\n>>> JUMP! (%"PRIu64" steps)\n\n", Njmp[jj]);
		
		RanJumpRan(Njmp[jj]);

		printstate();
		printf("\n");

		for (ii = 0; ii < 10; ii++)
		{
			printf("%"PRIu64"\n", pcg64dxsm_next());
		}
		starline();
		printf("\n");
	}
}
