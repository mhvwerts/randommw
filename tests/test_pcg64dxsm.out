
************************************************************
*** Seed : 0
************************************************************
state: 0x1aa1b5345996452d09585eb7a69561e3
inc:   0x418ddadb3af71a82588133bc447873a9

15672045205194312304
10230625629676741203
1393141542142426128
6186804329743392408
11731200791580184074
593908079603723752
16906657996318114573
4942292156362687787
6282811449723952844
11452040163060472522

state: 0x17dc31a80de6e163b3daf55d548203c9
inc:   0x418ddadb3af71a82588133bc447873a9

>>> JUMP! (7 steps)

state: 0x4493a8fce72ae07cfcb0c95dac91564c
inc:   0x418ddadb3af71a82588133bc447873a9

17482442208820442508
14144034442387100730
5538133234897647908
5051092106776430310
14613120975515737102
5332911046126341159
1679622300564794668
11311220931804488694
7745916272527462757
2366265781316940229
************************************************************


************************************************************
*** Seed : 12345
************************************************************
state: 0x1905e0335aae96349199b0d09775add5
inc:   0xc9c7353e6e2b1f287d761f2d4027fae7

17193872397121361007
6225879447261284483
4002610872796635837
6506281922641356830
10147648032342742849
16291563099318681940
16386928929934799094
5595252491649564630
8117175315690381675
6073746230066033075

state: 0x916eee226ea56a682ceb4ecdd373f90f
inc:   0xc9c7353e6e2b1f287d761f2d4027fae7

>>> JUMP! (64738 steps)

state: 0x4533ea27d1400e7437b28d6107eff3f9
inc:   0xc9c7353e6e2b1f287d761f2d4027fae7

9505272937505364404
6571929813102533142
13958594004192383150
4006906744314987319
12833941511434028622
16869381549270333770
14472796870589503523
12352716269376450856
14877903046021565760
12084178038737929955
************************************************************

